data/lua/wml-flow.lua

```diff
@@ -179,36 +179,51 @@ function wml_actions.foreach(cfg)
 	local array = wml.array_variables[array_name]
 	if #array == 0 then return end -- empty and scalars unwanted
 	local item_name = cfg.variable or "this_item"
-	local this_item <close> = utils.scoped_var(item_name) -- if this_item is already set
 	local i_name = cfg.index_var or "i"
-	local i <close> = utils.scoped_var(i_name) -- if i is already set
+
+	-- Some protection against unsupported modification of the array while iterating.
+	-- Changing the array will simply be ignored during the loop, and then overwritten
+	-- by the contents of "array" afterwards, this is simply a method of trying to add
+	-- an error message instead of silently ignoring those changes.
 	local array_length = wml.variables[array_name .. ".length"]
+	local check_for_modifications = true
+	if array_name:find("^" .. item_name) then
+		-- Disable the protection for WML such as array=this_item or array=this_item.abilities
+		check_for_modifications = false
+	end
 
-	for index, value in ipairs(array) do
-		-- Some protection against external modification
-		-- It's not perfect, though - it'd be nice if *any* change could be detected
-		if array_length ~= wml.variables[array_name .. ".length"] then
-			wml.error("WML array length changed during [foreach] iteration")
-		end
-		wml.variables[item_name] = value
-		-- set index variable
-		wml.variables[i_name] = index-1 -- here -1, because of WML array
-		-- perform actions
-		for do_child in wml.child_range(cfg, "do") do
-			local action = utils.handle_event_commands(do_child, "loop")
-			if action == "break" then
-				utils.set_exiting("none")
-				goto exit
-			elseif action == "continue" then
-				utils.set_exiting("none")
-				break
-			elseif action ~= "none" then
-				goto exit
+	do
+		local this_item <close> = utils.scoped_var(item_name)
+		local i <close> = utils.scoped_var(i_name)
+
+		for index, value in ipairs(array) do
+			do
```

Comment:
Why is there an extra scope here? It doesn't contain _any_ local variable, much less to-be-closed ones.

Reply:
Thanks for catching that, fixed.